### PR TITLE
Asynchronous transactions, Swift API

### PR DESCRIPTION
**Pull request by pavel-ship-it**
This pull request to enable asynchronous transactions in Realm Swift SDK.
The feature depends on [ObjC API](https://github.com/realm/realm-cocoa/pull/7551) and can be merged only after the mentioned PR.

```
    try? realm.writeAsync { _ in
        realm.create(SwiftStringObject.self, value: ["string"])
    } _: {
        // optional handling on write complete
    }

    try? realm.beginAsyncWrite { _ in
        realm.create(SwiftStringObject.self, value: ["string"])
        realm.commitAsyncWrite()
    }

    RLMAsyncTransactionId asyncTransactionId = try? realm.beginAsyncWrite { _ in
        // ...
    }
    try? realm.cancelAsyncWrite(asyncTransactionId)
```

### CODE REVIEW

File: RealmSwift/Realm.swift [hunk 1]
```diff
@@ -366,6 +370,80 @@ import Realm.Private
...
+     `block` should end by calling `commitAsyncWrite`, `cancelAsyncWrite`,
+     `commitWrite` or `cancelWrite`.
+     Returning without one of these calls will be equivalent to calling `commitAsyncWrite`.
+     Return Asynchronous transaction's handle.
+     @note `block` is queued for execution on the scheduler associated with the current realm.
+     It will run after the write mutex has been acquired.
+     The call returns immediately allowing the caller to proceed while the write mutex is held by someone else.
+     Write blocks from multiple calls to `beginAsyncWrite` or `writeAsync`
+     will be executed in order.
+     A later call to `beginAsyncWrite` or `writeAsync` will wait for any earlier
```
Comment (ericjordanmossman) [inline RealmSwift/Realm.swift:402]: Is this is a similar to way to say the previous sentence? 
"Write blocks from multiple calls to `beginAsyncWrite` or `writeAsync`
      will be executed in order."

Or are these two sentences trying to say different things?

File: RealmSwift/Realm.swift [hunk 2]
```diff
@@ -366,6 +370,80 @@ import Realm.Private
...
+     write blocks.
+     */
+    @discardableResult
+    public func beginAsyncWrite(_ asyncWriteBlock: @escaping (AsyncTransactionId) -> Void) throws -> AsyncTransactionId {
+        var asyncTransactionId: AsyncTransactionId = 0
```
Comment (ericjordanmossman) [inline RealmSwift/Realm.swift:407]: What does it mean for the asyncTransactionID to initialize as 0?

File: RealmSwift/Realm.swift [hunk 3]
```diff
@@ -366,6 +370,80 @@ import Realm.Private
...
+        }
+        return asyncTransactionId
+    }
+
+    /** Commit asynchronous transaction.
+     - parameter onComplete  is queued for execution on the scheduler associated with
+     the current realm. It will run after the commit has reached stable storage.
+     - parameter isGroupingAllowed     If `true`, the next `commitAsyncWrite` *may* run without an
+     intervening synchronization of stable storage.  Such a sequence of commits form a group.
+     It may help to have a better performance on write.
```
Comment (ericjordanmossman) [inline RealmSwift/Realm.swift:419]: I recommend changing `It` to `isGroupingAllowed`. So it's clearer that isGroupingAllowed may be performant.

File: RealmSwift/Realm.swift [hunk 4]
```diff
@@ -366,6 +370,80 @@ import Realm.Private
...
+    /** Cancel a queued code block (either for `writeAsync` or for`commitAsyncWrite`)
+     - note Cancelling a commit will not abort the commit, it will only cancel the callback
+     informing of commit completion.
+    */
+    public func cancelAsyncWrite(_  asyncTransactionId: AsyncTransactionId) throws {
+        rlmRealm.cancelAsyncTransaction(asyncTransactionId)
+    }
+
+    /**
+     Returns true when async transactiona has been created and the result of the last
```
Comment (ericjordanmossman) [inline RealmSwift/Realm.swift:438]: "transactions"?

File: RealmSwift/Tests/RealmTests.swift [hunk 5]
```diff
@@ -979,4 +979,413 @@ class RealmTests: TestCase {
...
+            }
+        })
+
+        try! realm.cancelAsyncWrite(asyncTransactionId)
+
+        waitForExpectations(timeout: 1, handler: nil)
+        XCTAssertNil(realm.objects(SwiftStringObject.self).first)
+    }
+
+    func testAsyncTransactionShouldCancelWithoutCommit() {
```
Comment (ericjordanmossman) [inline RealmSwift/Tests/RealmTests.swift:1036]: I'm not sure I understand this test.
What method call is cancelling the commit? What is causing the realm to still have no SwiftStringObjects?

Comment (ericjordanmossman) [reply RealmSwift/Tests/RealmTests.swift:1036]: Oh nvm. There is no commit. 🤦

File: RealmSwift/Tests/RealmTests.swift [hunk 6]
```diff
@@ -979,4 +979,413 @@ class RealmTests: TestCase {
...
+        realm.cancelWrite()
+
+        waitForExpectations(timeout: 1, handler: nil)
+        XCTAssertEqual(2, realm.objects(SwiftStringObject.self).count)
+    }
+
+    func testAsyncTransactionChangeNotification() {
+        let realm = try! Realm()
+        let asyncComplete = expectation(description: "async transaction complete")
+        asyncComplete.expectedFulfillmentCount = 4
```
Comment (ericjordanmossman) [inline RealmSwift/Tests/RealmTests.swift:1214]: Why is the .update case expected to be called 4 times when two writes are made?

File: RealmSwift/Realm.swift (hunk 2 above)
Comment (pavel-ship-it) [reply RealmSwift/Realm.swift:407]: it should be explicitly initialised in order to use it inside of escaping block

File: RealmSwift/Realm.swift (hunk 3 above)
Comment (pavel-ship-it) [reply RealmSwift/Realm.swift:419]: Nice

File: RealmSwift/Tests/RealmTests.swift (hunk 6 above)
Comment (pavel-ship-it) [reply RealmSwift/Tests/RealmTests.swift:1214]: 2 for write, 2 for update notification. I'll add another expectation for update to make it clear

File: RealmSwift/Realm.swift [hunk 7]
```diff
@@ -366,6 +370,115 @@ import Realm.Private
...
+     @param asyncWriteBlock The block containing actions to perform inside the write transaction.
+            `asyncWriteBlock` should end by calling `commitAsyncWrite` or `commitWrite`.
+            Returning without one of these calls is equivalent to calling `cancelAsyncWrite`.
+
+     @return An id identifying the asynchronous transaction which can be passed to
+             `cancelAsyncWrite` prior to the block being called to cancel
+             the pending invocation of the block.
+     */
+    @discardableResult
+    public func beginAsyncWrite(_ asyncWriteBlock: @escaping (AsyncTransactionId) -> Void) throws -> AsyncTransactionId {
```
Comment (tgoyne) [inline RealmSwift/Realm.swift:421]: This doesn't appear to ever actually throw an error and so shouldn't be `throws`.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,6 +1,24 @@
...
+    } _: { error in
+        // optional handling on write complete
+    }
+
+    try? realm.beginAsyncWrite { asyncTransactionId in
+        realm.create(SwiftStringObject.self, value: ["string"])
+        realm.commitAsyncWrite()
+    }
+
+    RLMAsyncTransactionId asyncTransactionId = try? realm.beginAsyncWrite { _ in
```
Comment (tgoyne) [inline CHANGELOG.md:17]:
```suggestion
    let asyncTransactionId = try? realm.beginAsyncWrite { _ in
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,6 +1,24 @@
 x.y.z Release notes (yyyy-MM-dd)
 =============================================================
 ### Enhancements
+* Add Swift API for asynchronous transactions
+```
+    try? realm.writeAsync { asyncTransactionId in
+        realm.create(SwiftStringObject.self, value: ["string"])
+    } _: { error in
```
Comment (tgoyne) [inline CHANGELOG.md:8]:
```suggestion
    } completion: { error in
```

File: RealmSwift/Realm.swift [hunk 10]
```diff
@@ -366,6 +369,112 @@ import Realm.Private
...
+
+     @param completionBlock A block which will be called on the source thread or queue
+                        once the commit has either completed or failed with an error.
+
+     @return An id identifying the asynchronous transaction which can be passed to
+             `cancelAsyncWrite` prior to the block being called to cancel
+             the pending invocation of the block.
+    */
+    @discardableResult
+    public func writeAsync(_ block: @escaping (AsyncTransactionId) -> Void, _ onComplete: ((Swift.Error?) -> Void)? = nil) -> AsyncTransactionId {
```
Comment (tgoyne) [inline RealmSwift/Realm.swift:391]: Completion callback should have an argument label.

Why does this and `beginAsyncWrite()` pass the transaction Id to the callback? The id is no longer usable for anything once the callback is invoked.

Comment (pavel-ship-it) [reply RealmSwift/Realm.swift:391]: Transaction Id is not required in `beginAsyncWrite` any more and can be removed. Thanks

File: Package.swift [hunk 11]
```diff
@@ -233,7 +233,8 @@ let package = Package(
                 "Realm/RLMUserAPIKey.mm"
             ],
             publicHeadersPath: "include",
-            cxxSettings: cxxSettings
+            cxxSettings: cxxSettings,
+            swiftSettings: testSwiftSettings
```
[the remark this answers is not in the document]
Comment (pavel-ship-it) [reply Package.swift:237]: those changes should gone together with feature flag. Fixed

**Review by ericjordanmossman — APPROVED** [review]

**Pull request merged**